sqlglot/optimizer/qualify_columns.py

```diff
@@ -380,18 +410,109 @@ def _qualify_columns(scope: Scope, resolver: Resolver) -> None:
                     column.set("table", column_table)
 
 
+def _expand_struct_stars(
+    expression: exp.Dot,
+    next_alias_name: t.Callable[[], str],
+) -> t.List[exp.Alias]:
+    """[BigQuery] Expand/Flatten foo.bar.* where bar is a struct column"""
+
+    dot_column = t.cast(exp.Column, expression.find(exp.Column))
+    if not dot_column.is_type(exp.DataType.Type.STRUCT):
+        return []
+
+    # All nested struct values are ColumnDefs, so normalize the first exp.Column in one
+    dot_column = dot_column.copy()
+    starting_struct = exp.ColumnDef(this=dot_column.this, kind=dot_column.type)
+
+    # First part is the table name and last part is the star so they can be dropped
+    dot_parts = expression.parts[1:-1]
+
+    # If we're expanding a nested struct (eg. t.c.f1.f2.*), start the DFS traversal from that struct
+    for part in dot_parts[1:]:
+        struct_fields = t.cast(exp.DataType, starting_struct.kind).expressions
+
+        if not struct_fields:
+            return []
+
+        for fld in struct_fields:
+            # Unable to expand star unless all struct fields are named
+            if not isinstance(fld.this, exp.Identifier):
+                return []
+
+            if fld.name == part.name and fld.kind.this == exp.DataType.Type.STRUCT:
+                starting_struct = fld
+
+    if starting_struct.name != dot_parts[-1].name:
+        return []
```

Comment:
When can this check fail?

Reply:
If the user defines an expansion such as `tbl.f0.f1.f2.*` but we're unable to find such fields in our first traversal (or we found them, but were unable to annotate them as structs).

This branch checks that the `starting_struct` is indeed the last part of the expression i.e `f2` in this case